projects/igniteui-angular/src/lib/grids/pivot-grid/pivot-util.ts

```diff
@@ -0,0 +1,476 @@
+import { cloneValue } from '../../core/utils';
+import { DataUtil } from '../../data-operations/data-util';
+import { FilteringLogic } from '../../data-operations/filtering-expression.interface';
+import { FilteringExpressionsTree } from '../../data-operations/filtering-expressions-tree';
+import { ISortingExpression } from '../../data-operations/sorting-strategy';
+import { IGridSortingStrategy, IgxSorting } from '../common/strategy';
+import { IPivotConfiguration, IPivotDimension, IPivotKeys, IPivotValue, PivotDimensionType } from './pivot-grid.interface';
+
+export class PivotUtil {
+    public static assignLevels(dims) {
+        for (const dim of dims) {
+            let currDim = dim;
+            let lvl = 0;
+            while (currDim.childLevel) {
+                currDim.level = lvl;
+                currDim = currDim.childLevel;
+                lvl++;
+            }
+            currDim.level = lvl;
+        }
+    }
+    public static getFieldsHierarchy(data: any[], dimensions: IPivotDimension[],
+        dimensionType: PivotDimensionType, pivotKeys: IPivotKeys): Map<string, any> {
+        const hierarchy = new Map<string, any>();
+        for (const rec of data) {
+            const vals = dimensionType === PivotDimensionType.Column ?
+                this.extractValuesForColumn(dimensions, rec, pivotKeys) :
+                this.extractValuesForRow(dimensions, rec, pivotKeys);
+            for (const [key, val] of vals) { // this should go in depth also vals.children
+                if (hierarchy.get(val.value) != null) {
+                    this.applyHierarchyChildren(hierarchy, val, rec, pivotKeys);
+                } else {
+                    hierarchy.set(val.value, cloneValue(val));
+                    this.applyHierarchyChildren(hierarchy, val, rec, pivotKeys);
+                }
+            }
+        }
+        return hierarchy;
+    }
+
+    public static sort(data: any[], expressions: ISortingExpression[], sorting: IGridSortingStrategy = new IgxSorting(), pivotKeys): any[] {
+        data.forEach(rec => {
+            const keys = Object.keys(rec);
+            rec.sorted = true;
+            keys.forEach(k => {
+                if (k.indexOf(pivotKeys.records) !== -1 && k !== pivotKeys.records) {
+                    const unsorted = rec[k].filter(x => !x.sorted);
+                    // sort all child record collections based on expression recursively.
+                    rec[k] = this.sort(unsorted, expressions, sorting, pivotKeys);
+                }
+            });
+        });
+        return DataUtil.sort(data, expressions, sorting);
+    }
+
+    public static extractValueFromDimension(dim: IPivotDimension, recData: any) {
+        return dim.memberFunction ? dim.memberFunction.call(null, recData) : recData[dim.memberName];
```

Comment:
I'm not sure if this comment is related only to the dimension sorting functionality or if you would also consider the pivot aggregations as something that should disregard casing. 

For example if you have 2 records:

```
{ "ProductCategory": "Clothing", ...},
{ "ProductCategory": "clothing", ... },
```
Does this imply that both should be part of the same aggregation group? So you get 1 "Clothing" row/column that aggregates both record values as opposed to 1 "Clothing" and 1 "clothing".

In general this is not the case with our existing pivot grids. When building groups for aggregation, they are case sensitive  by default- http://jsfiddle.net/gzh32onp/2/

Reply:
We discussed this with @rkaraivanov in the morning and igxGrid's grouping now would consider the two the same so  we'd like to have the same behavior by default for the PivotGrid too (should also be configurable, so that a user can instead use it as case-sensitive).

We can have it like it is right now if changing it is too problematic but give it a thought and we can discuss it further once you guys know how tough it will be.

igPivotGrid is probably not the best example because we don't know if it is consistent with igGrid grouping and sorting. Generally speaking these operations + aggregations are all based on grouping logic of same-ish values so that they should behave the same by default.

Reply:
User can set a custom memberFunction and can treat such values as same if he so wishes (for example .toLower() all values for that dimension). He has more control over the values compared to the igxGrid.